src/serialization/interval_log/mod.rs

```diff
@@ -0,0 +1,587 @@
+    /// recorded in this interval. If you're using a StartTime or BaseTime offset, you should
+    /// instead use a delta since that time. See the discussion about timestamps in the module-level
+    /// documentation.
+    ///
+    /// `duration` is the duration of the interval in seconds.
+    ///
+    /// `tag` is an optional tag for this histogram.
+    ///
+    /// `max_value_divisor` is used to scale down the max value to something that may be more human
+    /// readable. The max value in the log is only for human consumption, so you might prefer to
+    /// divide by 10<sup>9</sup> to turn nanoseconds into fractional seconds, for instance.
+    pub fn write_histogram<T: Counter>(
+        &mut self,
+        h: &Histogram<T>,
+        start_timestamp: f64,
+        duration: f64,
+        tag: Option<Tag>,
+        max_value_divisor: f64,
+    ) -> Result<(), IntervalLogWriterError<S::SerializeError>> {
+        self.internal_writer
+            .write_histogram(h, start_timestamp, duration, tag, max_value_divisor)
+    }
+}
+
+/// Errors that can occur while writing a log.
+#[derive(Debug)]
+pub enum IntervalLogWriterError<E> {
+    /// Histogram serialization failed.
+    SerializeError(E),
+    /// An i/o error occurred.
+    IoError(io::ErrorKind),
+}
+
+impl<E> From<io::Error> for IntervalLogWriterError<E> {
+    fn from(e: io::Error) -> Self {
+        IntervalLogWriterError::IoError(e.kind())
+    }
+}
+
+/// Write interval logs.
+struct InternalLogWriter<'a, 'b, W: 'a + io::Write, S: 'b + Serializer> {
+    writer: &'a mut W,
+    serializer: &'b mut S,
+    text_buf: String,
+    serialize_buf: Vec<u8>,
+}
+
+impl<'a, 'b, W: 'a + io::Write, S: 'b + Serializer> InternalLogWriter<'a, 'b, W, S> {
+    fn write_fmt(&mut self, args: fmt::Arguments) -> io::Result<()> {
+        self.writer.write_fmt(args)
+    }
+
+    fn write_comment(&mut self, s: &str) -> io::Result<()> {
+        write!(self.writer, "#{}\n", s)
+    }
+
+    fn write_histogram<T: Counter>(
+        &mut self,
+        h: &Histogram<T>,
+        start_timestamp: f64,
+        duration: f64,
+        tag: Option<Tag>,
+        max_value_divisor: f64,
+    ) -> Result<(), IntervalLogWriterError<S::SerializeError>> {
+        self.serialize_buf.clear();
+        self.text_buf.clear();
+
+        if let Some(Tag(s)) = tag {
+            write!(self.text_buf, "Tag={},", &s).expect("Writes to a String can't fail");
+        }
+
+        write!(
+            self.writer,
+            "{}{:.3},{:.3},{:.3},",
+            self.text_buf,
+            start_timestamp,
+            duration,
+            h.max() as f64 / max_value_divisor // because the Java impl does it this way
+        )?;
+
+        self.text_buf.clear();
+        let _len = self.serializer
+            .serialize(h, &mut self.serialize_buf)
+            .map_err(|e| IntervalLogWriterError::SerializeError(e))?;
```

Comment:
I would add
```rust
impl<E> From<E> for IntervalLogWriterError<E> {}
```
so that you can remove this `map_err`.

Reply:
Hmm... without a way to bound `E`, that would map every possible error type to `SerializeError`, which seems undesirable.

Reply:
But isn't that what the `IntervalLogWriterError` type already suggests?

Reply:
Well, there is the `IoError` variant, so it's not just a wrapper for serialization errors. Having everything else map to `SerializeError(E)` feels risky; wouldn't that make it possible to have a typo where some other `Result`-producing call gets a `?` appended to it, and presto now we're returning `SerializeError(SomeOtherUnrelatedError)`. 

`io::Error` seems like a reasonable thing for `From` because it's called many times and is unambiguous. I've never used an open-ended `From` like that so maybe my unease is unfounded, but it reminds me unpleasantly of Scala's implicits which made code hard to follow.

Reply:
What I mean is that `IntervalLogWriterError`'s one type argument is only used in the `SerializeError` variant. If we were to add some other wrapped error, that would be of a different type, which would change `IntervalLogWriterError` to take two type arguments `<E1, E2>`. In that case, an `impl` of `From` wouldn't make much sense without further bounds. But in the current case, if you have an error and want to produce some `IntervalLogWriterError<E>`, then that works exactly when that error is of type `E`. Which is what the above `From` would work for.

Reply:
I tightened up the bounds a bit:
```
pub enum IntervalLogWriterError<S: Serializer> {
    /// Histogram serialization failed.
    SerializeError(S::SerializeError),
    /// An i/o error occurred.
    IoError(io::ErrorKind),
}
```

Nonetheless, the compiler gets grumpy:
> error[E0119]: conflicting implementations of trait `std::convert::From<serialization::interval_log::IntervalLogWriterError<_>>` for type `serialization::interval_log::IntervalLogWriterError<_>

Reply:
What code it it's complaining on?

Reply:
I've extracted it into this playground: https://play.rust-lang.org/?gist=d6826495d99f9718c0980197a0fe14d3&version=stable

I think what the compiler is getting at is that you could have a `Serializer` whose `SerializeError` was `io::Error`, and then there would be no way to know which `impl` of `From` to use.

Reply:
Ah, yes, that makes sense. Oh well, leave it as-is I guess.